thoth/storages/graph/utils.py

```diff
@@ -0,0 +1,61 @@
+"""Various utilities and helper methods for the Graph database."""
+
+from aiogremlin.process.graph_traversal import AsyncGraphTraversalSource
+from gremlin_python.process.graph_traversal import addV
+from gremlin_python.process.graph_traversal import unfold
+
+from .models_base import VertexBase
+from .models_base import EdgeBase
+
+
+async def get_or_create_vertex(g: AsyncGraphTraversalSource, vertex: VertexBase) -> tuple:
+    """Create a vertex if not existed before, if the given vertex already exists, get tis id."""
+    query = g.V()
+    creation = addV(vertex.__label__)
+
+    for key, value in vertex.to_dict().items():
+        if value is not None:
+            query = query.has(key, value)
+            creation = creation.property(key, value)
+        else:
+            query = query.hasNot(key)
+
+    result = await query.fold().coalesce(
+            unfold().id().as_('id').constant(True).as_('existed').select('id', 'existed'),
+            creation.id().as_('id').constant(False).as_('existed').select('id', 'existed')
+        ).next()
+
+    # Assign to instance so instance has the id associated correctly for later queries.
+    vertex.id = result['id']
+
+    return result['id'], result['existed']
+
+
+async def get_or_create_edge(g: AsyncGraphTraversalSource, edge: EdgeBase) -> tuple:
+    """Create an edge if not existed before, if the given edge already exists, get its id."""
+    source_id = edge.source.id
+    target_id = edge.target.id
```

Comment:
note to myself: create an alternative that would accept directly ids so we can cache vertexes on worker/syncing side. The lookup on the caching side will be still O(N) considering variable properties, but N will be much smaller compared to N in the graph database (N = number of vertexes stored).